### PR TITLE
pull request #6

### PR DESCRIPTION
**Pull request by pkakelas**

### CODE REVIEW

File: controllers/image.php [hunk 1]
```diff
@@ -16,7 +16,7 @@ public static function create( $image ) {
                 $user->save();
             }
             catch ( ModelValidationException $e ) {
-                go( 'user', 'update', array( 'username' => $user->username, $e->error => true ) );
+                go( 'user', 'update', array( $e->error => true ) );
```
Comment (dionyziz) [inline controllers/image.php:19]: Why this change?

File: database/db_schema.sql [hunk 2]
```diff
@@ -28,10 +28,10 @@ SET time_zone = "+00:00";
 
 CREATE TABLE IF NOT EXISTS `countries` (
   `id` int(4) NOT NULL AUTO_INCREMENT,
-  `name` text COLLATE utf8mb4_unicode_ci NOT NULL,
-  `shortname` text COLLATE utf8mb4_unicode_ci NOT NULL,
+  `name` text COLLATE utf8_unicode_ci NOT NULL,
+  `shortname` text COLLATE utf8_unicode_ci NOT NULL,
   PRIMARY KEY (`id`)
-) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci ;
+) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci ;
```
Comment (dionyziz) [inline database/db_schema.sql]: This is already merged, please pull

File: database/db_schema.sql [hunk 3]
```diff
@@ -31,7 +31,46 @@ CREATE TABLE IF NOT EXISTS `countries` (
...
+-- --------------------------------------------------------
+
+--
+-- Table structure for table `gameusers`
+--
+
+CREATE TABLE IF NOT EXISTS `gameusers` (
+  `userid` int(11) NOT NULL,
+  `gameid` int(11) NOT NULL,
+  UNIQUE KEY `uc_gameusers` (`userid`,`gameid`)
```
Comment (dionyziz) [inline database/db_schema.sql]: Needs PRIMARY KEY, suggest order (gameid, userid)

File: database/db_schema.sql [hunk 4]
```diff
@@ -44,7 +83,24 @@ CREATE TABLE IF NOT EXISTS `images` (
...
+
+--
+-- Table structure for table `roundcreatures`
+--
+
+CREATE TABLE IF NOT EXISTS `roundcreatures` (
+  `roundid` int(11) NOT NULL,
+  `gameid` int(11) NOT NULL,
+  `creatureid` int(11) NOT NULL,
+  `desire` varchar(6) COLLATE utf8_unicode_ci NOT NULL,
```
Comment (dionyziz) [inline database/db_schema.sql]: Suggest ENUM type.

File: views/user/view.php [hunk 5]
```diff
@@ -6,6 +6,10 @@
             echo $user->image->target_path;
         ?>" alt="Profile Picture" width="100" height="100" /></p>
 
+<p><img src="<?php
+            echo $user->country->flag;
+        ?>" alt="Country flag" width="100" height="100" /></p>
```
Comment (dionyziz) [inline views/user/view.php]: Use alt text of actual country name.

**Comment by dionyziz** [comment]
I see no reason for the mode changes in .gitignore, .htaccess, README.md. Please revert these changes?

**Comment by dionyziz** [comment]
And LICENSE.

**Comment by dionyziz** [comment]
Seems like you've actually changed modes to all files in all directories. I don't see any reason for this. What is your rationale?

File: controllers/user.php [hunk 6]
```diff
@@ -96,7 +102,10 @@ public static function delete() {
         public static function createView( $username_empty, $username_invalid, $username_used, $email_empty, $email_used, $email_invalid, 
                 $password_empty, $password_not_matched, $password_small ) {
             include_once 'models/country.php'; 
+            include_once 'models/formtoken.php';
             $countries = Country::findAll();
+            $token = FormToken::create();
+            $_SESSION[ 'form' ][ 'token' ] = $token;
```
Comment (dionyziz) [inline controllers/user.php]: What happens if I have multiple tabs open? Let's say I open tab A, load a form, then open tab B, load a form, then go back to tab A and submit the form. The token check will fail. This is not something we want.

Tokens must persist across forms within the same session. One token per session.

File: database/db_schema.sql [hunk 7]
```diff
@@ -31,7 +31,46 @@ CREATE TABLE IF NOT EXISTS `countries` (
   `name` text COLLATE utf8_unicode_ci NOT NULL,
   `shortname` text COLLATE utf8_unicode_ci NOT NULL,
   PRIMARY KEY (`id`)
-) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci ;
+) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci AUTO_INCREMENT=1 ;
```
Comment (dionyziz) [inline database/db_schema.sql]: Please don't include AUTO_INCREMENT in the db schema.

File: database/db_schema.sql [hunk 8]
```diff
@@ -65,7 +121,7 @@ CREATE TABLE IF NOT EXISTS `users` (
   PRIMARY KEY (`id`),
   UNIQUE KEY `username` (`username`),
   UNIQUE KEY `email` (`email`)
-) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci ;
+) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci AUTO_INCREMENT=1 ;
```
Comment (dionyziz) [inline database/db_schema.sql]: similar

File: views/session/create.php [hunk 9]
```diff
@@ -26,6 +26,9 @@
             }
         ?>
         <p><input type="password" name="password" id="password" /></p>
+        <input type="hidden" name="token" value="<?php
```
Comment (dionyziz) [inline views/session/create.php]: Don't Repeat Yourself. We need a helper class for generating forms. It should automatically protect from forgery.

File: controllers/user.php [hunk 10]
```diff
@@ -14,6 +14,9 @@ public static function create( $username = '', $password = '', $password_repeat
             catch ( ModelNotFoundException $e ) {
                 $country = new Country();
             }
+            if ( !FormToken::validate( $token, $_SESSION[ 'form' ][ 'token' ] ) ) {
```
Comment (dionyziz) [inline controllers/user.php]: Don't Repeat Yourself. We can't rely on the developer writing this on every controller method that handles any form submission. Let's create a base controller class (BaseController), make every controller extend it, move dispatcher code from index.php to it, and at protection from forgery features to it.

File: views/user/view.php (hunk 5 above)
Comment (vitsalis) [inline views/user/view.php]: I fixed that

File: database/db_schema.sql (hunk 7 above)
Comment (vitsalis) [inline database/db_schema.sql]: fixed that

File: database/db_schema.sql (hunk 8 above)
Comment (vitsalis) [inline database/db_schema.sql]: fixed that

File: views/user/view.php (hunk 5 above)
Comment (dionyziz) [inline views/user/view.php]: OK I can't merge this unless Dimitris merges your fixes.

**Comment by dionyziz** [comment]
Dimitri, please also start using the present tense in your commit messages from now on:

"Add CSRF protection" instead of "added CSRF protection".

File: views/user/view.php (hunk 5 above)
Comment (vitsalis) [inline views/user/view.php]: Ok I told him to

Comment (dionyziz) [inline views/user/view.php]: You know you can always "pull request" from him.

**Comment by dionyziz** [comment]
There's nothing to merge.
